apps/emqx_gcp_device/src/emqx_gcp_device_api.erl

```diff
@@ -0,0 +1,459 @@
+    {ok, [validated_device()]} | {error, binary()}.
+validate_devices([], Validated) ->
+    {ok, lists:reverse(Validated)};
+validate_devices([Device | Devices], Validated) ->
+    case validate_device(Device) of
+        {ok, ValidatedDevice} ->
+            validate_devices(Devices, [ValidatedDevice | Validated]);
+        {error, _} = Error ->
+            Error
+    end.
+
+-spec validate_device(imported_device()) -> {ok, validated_device()} | {error, binary()}.
+validate_device(Device) ->
+    validate([deviceid, project, location, registry, keys, config], Device).
+
+-spec validate([checked_device_fields()], imported_device()) ->
+    {ok, validated_device()} | {error, binary()}.
+validate(Fields, Device) ->
+    validate(Fields, Device, #{}).
+
+-spec validate([checked_device_fields()], imported_device(), validated_device()) ->
+    {ok, validated_device()} | {error, binary()}.
+validate([], _Device, Validated) ->
+    {ok, Validated};
+validate([key_type | Fields], #{<<"key_type">> := KeyType} = Device, Validated) ->
+    case lists:member(KeyType, ?KEY_TYPES) of
+        true ->
+            validate(Fields, Device, Validated#{key_type => KeyType});
+        false ->
+            {error, <<"invalid key_type">>}
+    end;
+validate([key | Fields], #{<<"key">> := Key} = Device, Validated) ->
+    validate(Fields, Device, Validated#{key => Key});
+validate([expires_at | Fields], #{<<"expires_at">> := Expire} = Device, Validated) when
+    is_integer(Expire)
+->
+    validate(Fields, Device, Validated#{expires_at => Expire});
+validate([expires_at | _Fields], #{<<"expires_at">> := _}, _Validated) ->
+    {error, <<"invalid expires_at">>};
+validate([expires_at | Fields], Device, Validated) ->
+    validate(Fields, Device, Validated#{expires_at => 0});
+validate([Field | Fields], Device, Validated) when Field =:= deviceid; Field =:= key ->
+    FieldBin = atom_to_binary(Field),
+    case maps:find(FieldBin, Device) of
+        {ok, Value} when is_binary(Value) ->
+            validate(Fields, Device, Validated#{Field => Value});
+        _ ->
+            {error, <<"invalid or missing field: ", FieldBin/binary>>}
+    end;
+validate([Field | Fields], Device, Validated) when
+    Field =:= project; Field =:= location; Field =:= registry; Field =:= config
+->
+    FieldBin = atom_to_binary(Field),
+    case maps:find(FieldBin, Device) of
+        {ok, Value} when is_binary(Value) ->
+            validate(Fields, Device, Validated#{Field => Value});
+        error ->
+            validate(Fields, Device, Validated#{Field => <<>>});
+        _ ->
+            {error, <<"invalid field: ", FieldBin/binary>>}
+    end;
+validate([keys | Fields], #{<<"keys">> := Keys} = Device, Validated) when is_list(Keys) ->
+    try validate_keys(Keys) of
+        {ok, ValidatedKeys} ->
+            validate(Fields, Device, Validated#{keys => ValidatedKeys});
+        {error, _} = Error ->
+            Error
+    catch
+        error:badarg ->
```

Comment:
Q:
Where will throw this error, I can't find the location of it in either `validate_key` or `validate`

Reply:
I think that's not necessary. I'll remove this catch.